### PR TITLE
ARROW-14721: [C++] Strengthen DELTA_BYTE_ARRAY decoder

### PR DESCRIPTION
**Pull request by pitrou**
OSS-Fuzz found situations where invalid data could crash the DELTA_BYTE_ARRAY decoder.

Should fix the following issues:
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40865
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40902
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40922
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=41003

### CODE REVIEW

**Comment by pitrou** [comment]
@shanhuuang FYI

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -2401,24 +2413,31 @@ class DeltaByteArrayDecoder : public DecoderImpl,
...
       buffer[i].len += prefix_len_ptr[i];
       data_ptr += buffer[i].len;
-      last_value_ =
-          std::string(reinterpret_cast<const char*>(buffer[i].ptr), buffer[i].len);
+      prefix = string_view{reinterpret_cast<const char*>(buffer[i].ptr), buffer[i].len};
     }
     prefix_len_offset_ += max_values;
     this->num_values_ -= max_values;
     num_valid_values_ -= max_values;
+    last_value_ = prefix.to_string();
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:2440]: nit: to_string is a non-standard extension of our vendored version.  Is there a strong reason to use it here vs std::string(prefix)?

File: cpp/src/parquet/encoding.cc [hunk 2]
```diff
@@ -2265,21 +2269,27 @@ class DeltaLengthByteArrayDecoder : public DecoderImpl,
   }
 
   int Decode(ByteArray* buffer, int max_values) override {
+    // Decode up to `max_values` strings into an internal buffer
+    // and reference them into `buffer`.
     max_values = std::min(max_values, num_valid_values_);
 
-    int64_t data_size = 0;
+    int32_t data_size = 0;
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:2276]: it is not immediately clear to me why this must be less than int32_t?  I guess I need to reread the spec on this.

**Comment by emkornfield** [comment]
LGTM module a nit and a question about data sizing.

File: cpp/src/parquet/encoding.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/encoding.cc:2440]: No particular reason. I'll make the change.

File: cpp/src/parquet/encoding.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/parquet/encoding.cc:2276]: Not in the spec, but simply the fact that `GetBatch` takes an `int` for the batch size.

Comment (emkornfield) [reply cpp/src/parquet/encoding.cc:2276]: I think pages can only hold int32max values in them (row groups can hold more and apparently we have a bug someplace when reading).  So this makes sense.

**Pull request closed without merging**